### PR TITLE
Require partitioner.json.gz in VSM.read.

### PR DESCRIPTION
**Pull request by cseed**
Require partitioner.json.gz in VSM.read.

Added hc.write_partitioning to update legacy VDSes.

### CODE REVIEW

**Comment by cseed** [comment]
Hold off on merging this in so we can give people warning.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 1]
```diff
@@ -139,6 +139,26 @@ object VariantSampleMatrix {
       saSignature, vaSignature, globalSignature, wasSplit, isDosage)
   }
 
+  def writePartitioning(sqlContext: SQLContext, dirname: String): Unit = {
+    val sc = sqlContext.sparkContext
+    val hConf = sc.hadoopConfiguration
+
+    if (!hConf.exists(dirname + "/partitioner.json.gz"))
+      warn("write partitioning: partitioner.json.gz already exists, nothing to do")
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: return here.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 2]
```diff
@@ -139,6 +139,26 @@ object VariantSampleMatrix {
...
+
+    val parquetFile = dirname + "/rdd.parquet"
+
+    val fastKeys = sqlContext.readParquetSorted(parquetFile, Some(Array("variant")))
+      .map(_.getVariant(0))
+    val kvRDD = fastKeys.map(k => (k, ()))
+
+    val ordered = kvRDD.toOrderedRDD(fastKeys)
+
+    sqlContext.sparkContext.hadoopConfiguration.writeTextFile(dirname + "/partitioner.json.gz") { out =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: use hConf

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 3]
```diff
@@ -170,38 +190,16 @@ object VariantSampleMatrix {
...
-            OrderedPartitioner.read[Locus, Variant](in, rdd.partitions.length)
-          })
+          val jv = hConf.readFile(dirname + "/partitioner.json.gz")(JsonMethods.parse(_))
+          jv.fromJSON[OrderedPartitioner[Locus, Variant]]
         } catch {
-          case _: InvalidClassException => None
-          case _: ClassNotFoundException => None
-          case _: FileNotFoundException => None
+          case _: FileNotFoundException =>
+            fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: This is python-specific documentation in scala.  Feels a bit odd.

**Comment by cseed** [comment]
Back to you.

**Pull request merged**